macro-pdfviewer-api/src/main/java/com/xwiki/pdfviewer/internal/filter/MJSMimeTypeRegistrar.java

```diff
@@ -0,0 +1,175 @@
+            tikaConfig.getMimeRepository()
+                .addPattern(MimeTypes.getDefaultMimeTypes().forName(MediaType.application("javascript").toString()),
+                    "*.mjs");
+
+            // Create a new Tika instance with an associated mime type for the "*.mjs" pattern.
+            Tika newTika = new Tika(tikaConfig);
+            tikaField.set(null, newTika);
+            return true;
+        } catch (Exception e) {
+            logger.error("Failed to modify Tika field. Root cause: [{}]", ExceptionUtils.getRootCauseMessage(e));
+            return false;
+        }
+    }
+
+    private boolean uploadCustomMimeType()
+    {
+        ClassLoader classLoader = MimeTypesReader.class.getClassLoader();
+        String pathInJarWithJarPath =
+            classLoader.getResource(MimeTypesReader.class.getPackage().getName().replace('.', '/') + "/").getPath();
+        String jarFilePath =
+            URLDecoder.decode(pathInJarWithJarPath.substring(5, pathInJarWithJarPath.indexOf(EXCLAMATION_MARK)),
+                StandardCharsets.UTF_8);
+        String pathInJar =
+            pathInJarWithJarPath.substring(pathInJarWithJarPath.indexOf(EXCLAMATION_MARK) + 2) + CUSTOM_TYPE;
+        File tempJarFile = new File(jarFilePath + ".tmp");
+
+        try (JarFile jarFile = new JarFile(jarFilePath)) {
+            JarOutputStream jarOutputStream = new JarOutputStream(new FileOutputStream(tempJarFile));
+            // Copy the content of the tika jar file.
+            jarFile.stream().forEach(entry -> {
+                try {
+                    jarOutputStream.putNextEntry(new JarEntry(entry.getName()));
+                    if (!entry.isDirectory()) {
+                        InputStream entryInputStream = jarFile.getInputStream(entry);
+                        entryInputStream.transferTo(jarOutputStream);
+                    }
+                    jarOutputStream.closeEntry();
+                } catch (Exception e) {
+                    throw new RuntimeException(e);
+                }
+            });
+            // Add the custom-mimetypes.xml file.
+            JarEntry newEntry = new JarEntry(pathInJar);
+            jarOutputStream.putNextEntry(newEntry);
+            InputStream customFileInputStream =
+                MJSMimeTypeRegistrar.class.getClassLoader().getResourceAsStream(pathInJar);
+            customFileInputStream.transferTo(jarOutputStream);
+            jarOutputStream.closeEntry();
+        } catch (Exception e) {
+            if (tempJarFile.exists()) {
+                tempJarFile.delete();
+            }
+            logger.error("Failed to add the custom-mimetypes.xml to the tika jar file. Root cause: [{}]",
+                ExceptionUtils.getRootCauseMessage(e));
+            return false;
+        }
+
+        if (tempJarFile.exists()) {
+            File originalJar = new File(jarFilePath);
+            if (!originalJar.delete() || !tempJarFile.renameTo(originalJar)) {
```

Comment:
What if the delete operation succeeds but the rename fails? In any case all this feels too risky to me. I think it's enough to:

* try to modify Tika with reflection
* if this fails an exception is thrown and caught by the macro; the exception description should indicate that an administrator needs to put the ``custom-mimetypes.xml`` file in ``WEB-INF/classes/org/apache/tika/mime`` as indicated in the macro documentation

Reply:
Alright. I wanted to create a method to make it easier for an administrator to enable the correct mime type, and I don't think I can make it any safer than that. But you're right, as there can be unexpected errors as you mentioned: `delete operation succeeds but the rename fails`. I'll remove the runtime jar modification and only leave the reflection modification of Tika.